Image transforms library

amyeroberts opened the pull request:
# What does this PR do?

This is the first of a series of PRs to replace feature extractors with image processors for vision models. 

Create a new module `image_transforms.py` that will contain functions for transforming images e.g. `resize`. 

The functions are designed to: 
* Accept numpy arrays. 
* Return numpy arrays (except for e.g. `to_pil_image`)
* Provide logic such that the new image processors produce the same outputs as feature extractors when called directly. 

Subsequent PRs: 
* Image Processor Mixin: https://github.com/amyeroberts/transformers/pull/25
* GLPNImageProcessor: https://github.com/amyeroberts/transformers/pull/23
* GLPNFeatureExtractor -> GLPNImageProcessor alias https://github.com/amyeroberts/transformers/pull/24

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

amyeroberts committed:
Adapt FE methods to transforms library

amyeroberts committed:
Mixin for saving the image processor

amyeroberts committed:
Mixin for saving the image processor

amyeroberts committed:
Mixin for saving the image processor

amyeroberts committed:
Base processor skeleton

amyeroberts committed:
BatchFeature for packaging image processor outputs

amyeroberts committed:
BatchFeature for packaging image processor outputs

amyeroberts committed:
BatchFeature for packaging image processor outputs

amyeroberts committed:
Initial image processor for GLPN

amyeroberts committed:
REmove accidental import

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Fixup and docs

amyeroberts committed:
Import BatchFeature from feature_extraction_utils

amyeroberts committed:
Import BatchFeature from feature_extraction_utils

amyeroberts committed:
Import BatchFeature from feature_extraction_utils

amyeroberts committed:
Import BatchFeature from feature_extraction_utils

amyeroberts committed:
Merge branch 'image-processor-mixin' of github.com:amyeroberts/transformers into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Resolve conflicts

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Add rescale back and remove ImageType

amyeroberts committed:
fix import mistake

amyeroberts committed:
Fix enum var reference

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Can transform and specify image data format

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Remove redundant function

amyeroberts committed:
Update reference

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge in branch and remove conflicts

amyeroberts committed:
Add in rescaling

amyeroberts committed:
Data format flag for rescale

amyeroberts committed:
Fix typo

amyeroberts committed:
Fix dimension check

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Fixes to make IP and FE outputs match

amyeroberts committed:
Add tests for transforms

amyeroberts committed:
Add test for utils

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Update some docstrings

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Resole merge conflicts

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Make sure in channels last before converting to PIL

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Remove default to numpy batching

amyeroberts committed:
Fix up

amyeroberts committed:
Add docstring and model_input_types

amyeroberts committed:
Use feature processor config from hub

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Alias GLPN feature extractor to image processor

amyeroberts committed:
Alias feature extractor mixin

amyeroberts committed:
Resolve merge conflicts

amyeroberts committed:
Resolve merge conflicts

amyeroberts committed:
Resolve merge conflicts

amyeroberts committed:
Merge branch 'image-processor-glpn' into rename-fe-to-ip-glpn

amyeroberts committed:
Merge in main

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge and resolve conflicts

amyeroberts committed:
Merge branch 'image-processor-glpn' into rename-fe-to-ip-glpn

harrydrippin committed:
Fix TFSwinSelfAttention to have relative position index as non-trainable weight (#18226)

Signed-off-by: Seunghwan Hong <seunghwan@scatterlab.co.kr>

harrydrippin committed:
Refactor `TFSwinLayer` to increase serving compatibility (#18352)

* Refactor `TFSwinLayer` to increase serving compatibility

Signed-off-by: Seunghwan Hong <seunghwan@scatterlab.co.kr>

* Fix missed parameters while refactoring

Signed-off-by: Seunghwan Hong <seunghwan@scatterlab.co.kr>

* Fix window_reverse to calculate batch size

Signed-off-by: Seunghwan Hong <harrydrippin@gmail.com>
Co-Authored-By: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

ydshieh committed:
Add TF prefix to TF-Res test class (#18481)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Remove py.typed (#18485)

sgugger committed:
Fix pipeline tests (#18487)

* Fix pipeline tests

* Make sure all pipelines tests run with init changes

sgugger committed:
Use new huggingface_hub tools for download models (#18438)

* Draft new cached_file

* Initial draft for config and model

* Small fixes

* Fix first batch of tests

* Look in cache when internet is down

* Fix last tests

* Bad black, not fixing all quality errors

* Make diff less

* Implement change for TF and Flax models

* Add tokenizer and feature extractor

* For compatibility with main

* Add utils to move the cache and auto-do it at first use.

* Quality

* Deal with empty commit shas

* Deal with empty etag

* Address review comments

ydshieh committed:
Fix `test_dbmdz_english` by updating expected values (#18482)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Move cache folder to huggingface/hub for consistency with hf_hub (#18492)

* Move cache folder to just huggingface

* Thank you VsCode for this needless import

* Move to hub

* Forgot one

ydshieh committed:
Update some expected values in `quicktour.mdx` for `resampy 0.3.0` (#18484)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Forgot one new_ for cache migration

ydshieh committed:
disable Onnx test for google/long-t5-tglobal-base (#18454)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

julien-c committed:
Typo reported by Joel Grus on TWTR (#18493)

julien-c committed:
Just re-reading the whole doc every couple of months 😬 (#18489)

* Delete valohai.yaml

* NLP => ML

* typo

* website supports https

* datasets

* 60k + modalities

* unrelated link fixing for accelerate

* Ok those links were actually broken

* Fix link

* Make `AutoTokenizer` auto-link

* wording tweak

* add at least one non-nlp task

julien-c committed:
`transformers-cli login` => `huggingface-cli login` (#18490)

* zero chance anyone's using that constant no?

* `transformers-cli login` => `huggingface-cli login`

* `transformers-cli repo create` => `huggingface-cli repo create`

* `make style`

amyeroberts committed:
Add return_numpy=False flag for resize

amyeroberts committed:
Merge branch 'image-transforms-library' into image-processor-mixin

amyeroberts committed:
Merge branch 'image-processor-mixin' into base-image-processor-class

amyeroberts committed:
Merge branch 'base-image-processor-class' into image-batch-feature

amyeroberts committed:
Merge branch 'image-batch-feature' into image-processor-glpn

amyeroberts committed:
Merge branch 'image-processor-glpn' into rename-fe-to-ip-glpn

amyeroberts committed:
Fix up

amyeroberts committed:
Fix up

amyeroberts committed:
Use different frameworks safely

amyeroberts committed:
Safely import PIL

regisss committed:
Add seed setting to image classification example (#18519)

amyeroberts committed:
Call function checking if PIL available

Narsil committed:
[DX fix] Fixing QA pipeline streaming a dataset. (#18516)

* [DX fix] Fixing QA pipeline streaming a dataset.

QuestionAnsweringArgumentHandler would iterate over the whole dataset
effectively killing all properties of the pipeline.
This restores nice properties when using `Dataset` or `Generator` since
those are meant to be consumed lazily.

* Handling TF better.

sgugger committed:
Clean up hub (#18497)

* Clean up utils.hub

* Remove imports

* More fixes

* Last fix

pacman100 committed:
update fsdp docs (#18521)

* updating fsdp documentation

* typo fix

amyeroberts committed:
Only import if vision available

sgugger committed:
Fix compatibility with 1.12 (#17925)

* Fix compatibility with 1.12

* Remove pin from examples requirements

* Update torch scatter version

* Fix compatibility with 1.12

* Remove pin from examples requirements

* Update torch scatter version

* fix torch.onnx.symbolic_opset12 import

* Reject bad version

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

sgugger committed:
Remove debug statement

ankrgyl committed:
Specify en in doc-builder README example (#18526)

Co-authored-by: Ankur Goyal <ankur@impira.com>

sgugger committed:
New cache fixes: add safeguard before looking in folders (#18522)

amyeroberts commented:
@sgugger @NielsRogge @alaradirik @LysandreJik  Adding you all for a first-pass review for the draft ImageProcessor work. This PR is failing because it's not safely importing e.g. `PIL` if it's not available, but the core logic shouldn't change. I'll add you to the follow up PRs too. Note: `ImageProcessor` has only been implemented for the GLPN model so far.

sgugger commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,216 @@
+        # PIL expects image to have channels last
+        image = to_channel_dimension_format(image, ChannelDimension.LAST)
+        image = to_pil_image(image)
+    h, w = size
```
Use height, width here
```suggestion
    height, width = size
```

sgugger commented on src/transformers/image_utils.py:
```diff
@@ -14,33 +14,98 @@
+    elif image.ndim == 4:
+        first_dim, last_dim = 1, 3
+    else:
+        raise ValueError("Unsupported number of image dimensions: {}".format(image.ndim))
```
We use f-strings everywhere in the library :-)

sgugger reviewed (commented):
Core, looks good to me, just left a couple of comments.

ydshieh committed:
unpin resampy (#18527)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

stevhliu committed:
✨ update to use interlibrary links instead of Markdown (#18500)

stevhliu committed:
Add example of multimodal usage to pipeline tutorial (#18498)

* 📝 add example of multimodal usage to pipeline tutorial

* 🖍 apply feedbacks

* 🖍 apply niels feedback

NielsRogge committed:
[VideoMAE] Add model to doc tests (#18523)

* Add videomae to doc tests

* Add pip install decord

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

mishig25 committed:
Update perf_train_gpu_one.mdx (#18532)

Rasmusafj committed:
Update no_trainer.py scripts to include accelerate gradient accumulation wrapper (#18473)

* Added accelerate gradient accumulation wrapper to run_image_classification_no_trainer.py example script

* make fixup changes

* PR comments

* changed input to Acceletor based on PR comment, ran make fixup

* Added comment explaining the sync_gradients statement

* Fixed lr scheduler max steps

* Changed run_clm_no_trainer.py script to use accelerate gradient accum wrapper

* Fixed all scripts except wav2vec2 pretraining to use accelerate gradient accum wrapper

* Added accelerate gradient accum wrapper for wav2vec2_pretraining_no_trainer.py script

* make fixup and lr_scheduler step inserted back into run_qa_beam_search_no_trainer.py

* removed changes to run_wav2vec2_pretraining_no_trainer.py script and fixed using wrong constant in qa_beam_search_no_trainer.py script

donelianc committed:
Add Spanish translation of converting_tensorflow_models.mdx (#18512)

* Add file in spanish docs to be translated

* Finish translation to Spanish

* Improve Spanish  wording

* Add suggested changes from review

AguilaCudicio committed:
Spanish translation of summarization.mdx (#15947) (#18477)

* Add Spanish translation of summarization.mdx

* Apply suggestions from code review

Co-authored-by: Omar U. Espejel <espejelomar@gmail.com>

Co-authored-by: Omar U. Espejel <espejelomar@gmail.com>

younesbelkada committed:
Let's not cast them all (#18471)

* add correct dtypes when checking for params dtype

* forward contrib credits

* Update src/transformers/modeling_utils.py

Co-authored-by: Thomas Wang <24695242+thomasw21@users.noreply.github.com>

* more comments

- added more comments on why we cast only floating point parameters

* Update src/transformers/modeling_utils.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Co-authored-by: sgugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Thomas Wang <24695242+thomasw21@users.noreply.github.com>

NikeNano committed:
fix: data2vec-vision Onnx ready-made configuration. (#18427)

* feat: add the data2vec conf that are missing https://huggingface.co/docs/transformers/serialization

* fix: wrong config

chainyo committed:
Add mt5 onnx config (#18394)

* update features

* MT5OnnxConfig added with updated with tests and docs

* fix imports

* fix onnc_config_cls for mt5

Co-authored-by: Thomas Chaigneau <thomas.deeptools.ai>

ydshieh committed:
Minor update of `run_call_with_unpacked_inputs` (#18541)

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

younesbelkada committed:
BART - Fix attention mask device issue on copied models (#18540)

* attempt to fix attn mask device

* fix bart `_prepare_decoder_attention_mask`

- add correct device
- run `make fix-copies` to propagate the fix

Narsil committed:
Adding a new `align_to_words` param to qa pipeline. (#18010)

* Adding a new `align_to_words` param to qa pipeline.

* Update src/transformers/pipelines/question_answering.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Import protection.

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stevhliu committed:
📝 update metric with evaluate (#18535)

amyeroberts committed:
Address Sylvain PR comments
Co-authored-by: Sylvain.gugger@gmail.com

YouJiacheng committed:
Restore _init_weights value in no_init_weights (#18504)

* Recover _init_weights value in no_init_weights

For potential nested use. 
In addition, users might modify private no_init_weights as well.

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Remove private variable change check

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

sgugger committed:
Clean up comment

stevhliu committed:
📝 update documentation build section (#18548)

younesbelkada committed:
`bitsandbytes` - `Linear8bitLt` integration into `transformers` models (#17901)

* first commit

* correct replace function

* add final changes

- works like charm!
- cannot implement tests yet
- tested

* clean up a bit

* add bitsandbytes dependencies

* working version

- added import function
- added bitsandbytes utils file

* small fix

* small fix

- fix import issue

* fix import issues

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* refactor a bit

- move bitsandbytes utils to utils
- change comments on functions

* reformat docstring

- reformat docstring on init_empty_weights_8bit

* Update src/transformers/__init__.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* revert bad formatting

* change to bitsandbytes

* refactor a bit

- remove init8bit since it is useless

* more refactoring

- fixed init empty weights issue
- added threshold param

* small hack to make it work

* Update src/transformers/modeling_utils.py

* Update src/transformers/modeling_utils.py

* revmoe the small hack

* modify utils file

* make style + refactor a bit

* create correctly device map

* add correct dtype for device map creation

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* apply suggestions

- remove with torch.grad
- do not rely on Python bool magic!

* add docstring

 - add docstring for new kwargs

* add docstring

- comment `replace_8bit_linear` function
- fix weird formatting

* - added more documentation
- added new utility function for memory footprint tracking
- colab demo to add

* few modifs

- typo doc
- force cast into float16 when load_in_8bit is enabled

* added colab link

* add test architecture + docstring a bit

* refactor a bit testing class

* make style + refactor a bit

* enhance checks

- add more checks
- start writing saving test

* clean up a bit

* male style

* add more details on doc

* add more tests

- still needs to fix 2 tests

* replace by "or"

- could not fix it from GitHub GUI

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* refactor a bit testing code + add readme

* make style

* fix import issue

* Update src/transformers/modeling_utils.py

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* add few comments

* add more doctring + make style

* more docstring

* raise error when loaded in 8bit

* make style

* add warning if loaded on CPU

* add small sanity check

* fix small comment

* add bitsandbytes on dockerfile

* Improve documentation

- improve documentation from comments

* add few comments

* slow tests pass on the VM but not on the CI VM

* Fix merge conflict

* make style

* another test should pass on a multi gpu setup

* fix bad import in testing file

* Fix slow tests

- remove dummy batches
- no more CUDA illegal memory errors

* odify dockerfile

* Update docs/source/en/main_classes/model.mdx

* Update Dockerfile

* Update model.mdx

* Update Dockerfile

* Apply suggestions from code review

* few modifications

- lm head can stay on disk/cpu
- change model name so that test pass

* change test value

- change test value to the correct output
- torch bmm changed to baddmm in bloom modeling when merging

* modify installation guidelines

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* replace `n`by `name`

* merge `load_in_8bit` and `low_cpu_mem_usage`

* first try - keep the lm head in full precision

* better check

- check the attribute `base_model_prefix` instead of computing the number of parameters

* added more tests

* Update src/transformers/utils/bitsandbytes.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Merge branch 'integration-8bit' of https://github.com/younesbelkada/transformers into integration-8bit

* improve documentation

- fix typos for installation
- change title in the documentation

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

gante committed:
TF: XLA-trainable DeBERTa v2 (#18546)

* fix deberta issues

* add different code paths for gpu and tpu

* shorter gpu take along axis

* Stable Dropout without tf cond

* variable must be float

sgugger committed:
Preserve hub-related kwargs in AutoModel.from_pretrained (#18545)

* Preserve hub-related kwargs in AutoModel.from_pretrained

* Fix tests

* Remove debug statement

amyeroberts committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <Sylvain.gugger@gmail.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

Rocketknight1 committed:
TF Examples Rewrite (#18451)

* Finished QA example

* Dodge a merge conflict

* Update text classification and LM examples

* Update NER example

* New Keras metrics WIP, fix NER example

* Update NER example

* Update MC, summarization and translation examples

* Add XLA warnings when shapes are variable

* Make sure batch_size is consistently scaled by num_replicas

* Add PushToHubCallback to all models

* Add docs links for KerasMetricCallback

* Add docs links for prepare_tf_dataset and jit_compile

* Correct inferred model names

* Don't assume the dataset has 'lang'

* Don't assume the dataset has 'lang'

* Write metrics in text classification

* Add 'framework' to TrainingArguments and TFTrainingArguments

* Export metrics in all examples and add tests

* Fix training args for Flax

* Update command line args for translation test

* make fixup

* Fix accidentally running other tests in fp16

* Remove do_train/do_eval from run_clm.py

* Remove do_train/do_eval from run_mlm.py

* Add tensorflow tests to circleci

* Fix circleci

* Update examples/tensorflow/language-modeling/run_mlm.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/test_tensorflow_examples.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/translation/run_translation.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Update examples/tensorflow/token-classification/run_ner.py

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

* Fix save path for tests

* Fix some model card kwargs

* Explain the magical -1000

* Actually enable tests this time

* Skip text classification PR until we fix shape inference

* make fixup

Co-authored-by: Joao Gante <joaofranciscocardosogante@gmail.com>

sgugger committed:
Use commit hash to look in cache instead of calling head (#18534)

* Use commit hash to look in cache instead of calling head

* Add tests

* Add attr for local configs too

* Stupid typos

* Fix tests

* Update src/transformers/utils/hub.py

Co-authored-by: Julien Chaumond <julien@huggingface.co>

* Address Julien's comments

Co-authored-by: Julien Chaumond <julien@huggingface.co>

julien-c committed:
`pipeline` support for `device="mps"` (or any other string) (#18494)

* `pipeline` support for `device="mps"` (or any other string)

* Simplify `if` nesting

* Update src/transformers/pipelines/base.py

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Fix? @sgugger

* passing `attr=None` is not the same as not passing `attr` 🤯

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stevhliu committed:
Update philosophy to include other preprocessing classes (#18550)

* 📝 update philosophy to include other preprocessing classes

* 🖍 apply feedbacks

sgugger committed:
Properly move cache when it is not in default path (#18563)

unography committed:
Adds CLIP to models exportable with ONNX (#18515)

* onnx config for clip

* default opset as 14

* changes from the original repo

* input values order fix

* outputs fix

* remove unused import

* ran make fix-copies

* black format

* review comments: forward ref, import fix, model change revert, .to cleanup

* make style

* formatting fixes

* revert groupvit

* comment for cast to int32

* comment fix

* make .T as .t() for onnx conversion

* ran make fix-copies

* remove unneeded comment

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* fix copies

* remove comment

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

ydshieh committed:
raise atol for MT5OnnxConfig (#18560)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

mrwyattii committed:
fix string (#18568)

joihn committed:
Segformer TF: fix output size in documentation (#18572)

* Segformer TF: fix output size in doc

* Segformer pytorch: fix output size in doc

Co-authored-by: Maxime Gardoni <maxime.gardoni@ecorobotix.com>

alaradirik committed:
Fix resizing bug in OWL-ViT (#18573)

* Fixes resizing bug in OWL-ViT
* Defaults to square resize if size is set to an int
* Sets do_center_crop default value to False

pocca2048 committed:
Fix LayoutLMv3 documentation (#17932)

* fix typos

* fix sequence_length docs of LayoutLMv3Model

* delete trailing white spaces

* fix layoutlmv3 docs more

* apply make fixup & quality

* change to two versions of input docstring

* apply make fixup & quality

alaradirik commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,216 @@
+    return PIL.Image.fromarray(image)
+
+
+def get_resize_output_image_size(
```
The draft PR looks good to me! I have a question though, do we need two methods for resizing (`get_resize_output_image_size` and `resize`)? 

The `default_to_square` argument is useful and I think removing it from the `resize` method might break some pipelines

amyeroberts replied:
The two main reasons for having both `get_resize_output_image_size` and `resize` were:
* Separation of concerns - `resize` is only responsible for resizing an image, not shape calculations. Coming from TF I found the `resize` behaviour surprising. This forces a more explicit shape passing when processing. 
* The shape logic is quite complex, this broke it down into smaller functions. If the shape calculation were moved back inside `resize` I would probably still have put the logic outside in a private function, but this is a style preference. 

I don't feel very strongly about this and happy to move the finding shape logic to be internal to the method again. 

For `default_to_square` where do you think this might break? Now you've highlighted - I could foresee in cases when individual feature extractor methods are used, although I haven't seen this any of our pipelines or notebooks. Let me know if there's something I'm overlooking or examples.

alaradirik reviewed (commented):
I think the first draft looks great! 

I just have a comment on the `resize` method changes. I'm in favor of keeping it as it is as some of the feature extractors call the resize method with a int value for square resizing and this might break some things later on

sgugger committed:
Skip broken tests

donebydan committed:
Change BartLearnedPositionalEmbedding's forward method signature to support Opacus training (#18486)

* changing BartLearnedPositionalEmbedding forward signature and references to it

* removing debugging dead code (thanks style checker)

* blackened modeling_bart file

* removing copy inconsistencies via make fix-copies

* changing references to copied signatures in Bart variants

* make fix-copies once more

* using expand over repeat (thanks @michaelbenayoun)

* expand instead of repeat for all model copies

Co-authored-by: Daniel Jones <jonesdaniel@microsoft.com>

flozi00 committed:
german docs translation (#18544)

* Create _config.py

* Create _toctree.yml

* Create index.mdx

not sure about "du / ihr" oder "sie"

* Create quicktour.mdx

* Update _toctree.yml

* Update build_documentation.yml

* Update build_pr_documentation.yml

* fix build

* Update index.mdx

* Update quicktour.mdx

* Create installation.mdx

* Update _toctree.yml

iiLaurens committed:
Deberta V2: Fix critical trace warnings to allow ONNX export (#18272)

* Fix critical trace warnings to allow ONNX export

* Force input to `sqrt` to be float type

* Cleanup code

* Remove unused import statement

* Update model sew

* Small refactor

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* Use broadcasting instead of repeat

* Implement suggestion

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

* Match deberta v2 changes in sew_d

* Improve code quality

* Update code quality

* Consistency of small refactor

* Match changes in sew_d

Co-authored-by: Michael Benayoun <mickbenayoun@gmail.com>

michaelbenayoun committed:
[FX] _generate_dummy_input supports audio-classification models for labels (#18580)

* Support audio classification architectures for labels generation, as well as provides a flag to print warnings or not

* Use ENV_VARS_TRUE_VALUES

sgugger committed:
Fix docstrings with last version of hf-doc-builder styler (#18581)

* Fix docstrings with last version of hf-doc-builder styler

* Remove empty Parameter block

dependabot[bot] committed:
Bump nbconvert from 6.0.1 to 6.3.0 in /examples/research_projects/lxmert (#18565)

Bumps [nbconvert](https://github.com/jupyter/nbconvert) from 6.0.1 to 6.3.0.
- [Release notes](https://github.com/jupyter/nbconvert/releases)
- [Commits](https://github.com/jupyter/nbconvert/compare/6.0.1...6.3.0)

---
updated-dependencies:
- dependency-name: nbconvert
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump nbconvert in /examples/research_projects/visual_bert (#18566)

Bumps [nbconvert](https://github.com/jupyter/nbconvert) from 6.0.1 to 6.3.0.
- [Release notes](https://github.com/jupyter/nbconvert/releases)
- [Commits](https://github.com/jupyter/nbconvert/compare/6.0.1...6.3.0)

---
updated-dependencies:
- dependency-name: nbconvert
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

alaradirik committed:
fix owlvit tests, update docstring examples (#18586)

amyeroberts committed:
Return the permuted hidden states if return_dict=True (#18578)

ArthurZucker committed:
Load sharded pt to flax (#18419)

* initial commit

* add small test

* add cross pt tf flag to test

* fix quality

* style

* update test with new repo

* fix failing test

* update

* fix wrong param ordering

* style

* update based on review

* update related to recent new caching mechanism

* quality

* Update based on review

Co-authored-by: sgugger <sylvain.gugger@gmail.com>

* quality and style

* Update src/transformers/modeling_flax_utils.py
Co-authored-by: sgugger <sylvain.gugger@gmail.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

amyeroberts committed:
Update src/transformers/image_transforms.py

Co-authored-by: Alara Dirik <8944735+alaradirik@users.noreply.github.com>

donelianc committed:
Add type hints for ViLT models (#18577)

* Add type hints for Vilt models

* Add missing return type for TokenClassification class

sywangyi committed:
update doc for perf_train_cpu_many, add intel mpi introduction (#18576)

* update doc for perf_train_cpu_many, add mpi introduction

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

* Update docs/source/en/perf_train_cpu_many.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* Update docs/source/en/perf_train_cpu_many.mdx

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>

Signed-off-by: Wang, Yi A <yi.a.wang@intel.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

stas00 committed:
typos (#18594)

pacman100 committed:
FSDP bug fix for `load_state_dict` (#18596)

ydshieh committed:
Add `TFAutoModelForSemanticSegmentation` to the main `__init__.py` (#18600)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

gante committed:
Generate: validate `model_kwargs` (and catch typos in generate arguments) (#18261)

* validate generate model_kwargs

* generate tests -- not all models have an attn mask

younesbelkada committed:
Supporting seq2seq models for `bitsandbytes` integration (#18579)

* Supporting seq2seq models for `bitsandbytes` integration

- `bitsandbytes` integration supports now seq2seq models
- check if a model has tied weights as an additional check

* small modification

- tie the weights before looking at tied weights!

NielsRogge committed:
Add Donut (#18488)

* First draft

* Improve script

* Update script

* Make conversion work

* Add final_layer_norm attribute to Swin's config

* Add DonutProcessor

* Convert more models

* Improve feature extractor and convert base models

* Fix bug

* Improve integration tests

* Improve integration tests and add model to README

* Add doc test

* Add feature extractor to docs

* Fix integration tests

* Remove register_buffer

* Fix toctree and add missing attribute

* Add DonutSwin

* Make conversion script work

* Improve conversion script

* Address comment

* Fix bug

* Fix another bug

* Remove deprecated method from docs

* Make Swin and Swinv2 untouched

* Fix code examples

* Fix processor

* Update model_type to donut-swin

* Add feature extractor tests, add token2json method, improve feature extractor

* Fix failing tests, remove integration test

* Add do_thumbnail for consistency

* Improve code examples

* Add code example for document parsing

* Add DonutSwin to MODEL_NAMES_MAPPING

* Add model to appropriate place in toctree

* Update namespace to appropriate organization

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

amyeroberts committed:
Update src/transformers/models/glpn/feature_extraction_glpn.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

NielsRogge committed:
Fix URLs (#18604)

Co-authored-by: Niels Rogge <nielsrogge@Nielss-MacBook-Pro.local>

Muennighoff committed:
Update BLOOM parameter counts (#18531)

* Update BLOOM parameter counts

* Update BLOOM parameter counts

stas00 committed:
[doc] fix anchors (#18591)

the manual anchors end up being duplicated with automatically added anchors and no longer work.

stas00 committed:
[fsmt] deal with -100 indices in decoder ids (#18592)

* [fsmt] deal with -100 indices in decoder ids

Fixes: https://github.com/huggingface/transformers/issues/17945

decoder ids get the default index -100, which breaks the model - like t5 and many other models add a fix to replace -100 with the correct pad index. 

For some reason this use case hasn't been used with this model until recently - so this issue was there since the beginning it seems.

Any suggestions to how to add a simple test here? or perhaps we have something similar already? user's script is quite massive.

* style

younesbelkada committed:
small change (#18584)

KMFODA committed:
Flax Remat for LongT5 (#17994)

* [Flax] Add remat (gradient checkpointing)

* fix variable naming in test

* flip: checkpoint using a method

* fix naming

* fix class naming

* apply PVP's suggestions from code review

* add gradient_checkpointing to examples

* Add gradient_checkpointing to run_mlm_flax

* Add remat to longt5

* Add gradient checkpointing test longt5

* Fix args errors

* Fix remaining tests

* Make fixup & quality fixes

* replace kwargs

* remove unecessary kwargs

* Make fixup changes

* revert long_t5_flax changes

* Remove return_dict and copy to LongT5

* Remove test_gradient_checkpointing

Co-authored-by: sanchit-gandhi <sanchit@huggingface.co>

pacman100 committed:
mac m1 `mps` integration (#18598)

* mac m1 `mps` integration

* Update docs/source/en/main_classes/trainer.mdx

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

* addressing comments

* Apply suggestions from code review

Co-authored-by: Dan Saattrup Nielsen <47701536+saattrupdan@users.noreply.github.com>

* resolve comment

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: Dan Saattrup Nielsen <47701536+saattrupdan@users.noreply.github.com>

ydshieh committed:
Change scheduled CIs to use torch 1.12.1 (#18644)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

ydshieh committed:
Add checks for some workflow jobs (#18583)

Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

gante committed:
TF: Fix generation repetition penalty with XLA (#18648)

flozi00 committed:
Update longt5.mdx (#18634)

amyeroberts committed:
Merge pull request #25 from amyeroberts/image-processor-mixin

Image processor mixin

zhoutang776 committed:
Update run_translation_no_trainer.py (#18637)

* Update run_translation_no_trainer.py

found an error in selecting `no_decay` parameters and some small modifications when the user continues to train from a checkpoint

* fixs `no_decay` and `resume_step` issue

1. change `no_decay` list
2. if use continue to train their model from provided checkpoint, the `resume_step` will not be initialized properly if `args.gradient_accumulation_steps != 1`

younesbelkada committed:
[bnb] Minor modifications (#18631)

* bnb minor modifications

- refactor documentation
- add troubleshooting README
- add PyPi library on DockerFile

* Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Apply suggestions from code review

* Apply suggestions from code review

* Apply suggestions from code review

* put in one block

- put bash instructions in one block

* update readme

- refactor a bit hardware requirements

* change text a bit

* Apply suggestions from code review

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

* apply suggestions

Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

* add link to paper

* Apply suggestions from code review

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* Update tests/mixed_int8/README.md

* Apply suggestions from code review

* refactor a bit

* add instructions Turing & Amperer

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>

* add A6000

* clarify a bit

* remove small part

* Update tests/mixed_int8/README.md

Co-authored-by: Stas Bekman <stas00@users.noreply.github.com>
Co-authored-by: Yih-Dar <2521628+ydshieh@users.noreply.github.com>

stefan-it committed:
Examples: add Bloom support for token classification (#18632)

* examples: add Bloom support for token classification (FLAX, PyTorch and TensorFlow)

* examples: remove support for Bloom in token classication (FLAX and TensorFlow currently have no support for it)

ydshieh committed:
Fix Yolos ONNX export test (#18606)

Co-authored-by: lewtun <lewis.c.tunstall@gmail.com>
Co-authored-by: ydshieh <ydshieh@users.noreply.github.com>

amyeroberts committed:
Add in docstrings

amyeroberts committed:
Merge pull request #23 from amyeroberts/image-processor-glpn

Image processor glpn

amyeroberts committed:
Merge pull request #26 from amyeroberts/image-batch-feature

BaseImageProcessor

amyeroberts committed:
Merge pull request #24 from amyeroberts/rename-fe-to-ip-glpn

GLPN - Rename featue extractor to image processor glpn

amyeroberts committed:
Fixup

amyeroberts committed:
Fix up

amyeroberts committed:
Resolve conflicts

amyeroberts committed:
Move PIL default arguments inside function for safe imports

amyeroberts committed:
Add image utils to toctree

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_18520). All of your documentation changes will be reflected on that endpoint.

amyeroberts commented on src/transformers/image_utils.py:
```diff
@@ -220,7 +315,7 @@ def normalize(self, image, mean, std):
             return (image - mean) / std
 
-    def resize(self, image, size, resample=PIL.Image.BILINEAR, default_to_square=True, max_size=None):
+    def resize(self, image, size, resample=None, default_to_square=True, max_size=None):
```
Moved to be within the function, rather than in the signature to avoid issues when importing and PIL isn't installed

amyeroberts commented on src/transformers/image_utils.py:
```diff
@@ -377,7 +474,7 @@ def flip_channel_order(self, image):
         return image[::-1, :, :]
 
-    def rotate(self, image, angle, resample=PIL.Image.NEAREST, expand=0, center=None, translate=None, fillcolor=None):
+    def rotate(self, image, angle, resample=None, expand=0, center=None, translate=None, fillcolor=None):
```
Moved to be within the function, rather than in the signature to avoid issues when importing and PIL isn't installed

amyeroberts commented:
@alaradirik @sgugger I've now merged in the stacked PRs above this one. This PR has the transforms library and the image processor for GLPN. Thanks for all of you reviews so far! This should be ready for a final review to make sure all the pieces work together before merging.

amyeroberts committed:
Update `rescale` method to reflect changes in #18677

NielsRogge commented on docs/source/en/internal/image_processing_utils.mdx:
```diff
@@ -0,0 +1,32 @@
+
+[[autodoc]] image_transforms.to_pil_image
+
+
+
```
```suggestion
```

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+
+    Args:
+        image (`numpy.ndarray`):
+            The image to convert to the PIL Image format.
```
This docstring doesn't seem right

amyeroberts replied:
Yep, sorry, I'll fix

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    if current_channel_dim == target_channel_dim:
+        return image
+
+    if target_channel_dim == ChannelDimension.FIRST:
+        return image.transpose((2, 0, 1))
+
+    if target_channel_dim == ChannelDimension.LAST:
+        return image.transpose((1, 2, 0))
+
+    raise ValueError("Unsupported channel dimension format: {}".format(channel_dim))
```
Wouldn't this be cleaner with an if, elif, else?

amyeroberts replied:
It depends who you ask and what you mean by clean :D 

I picked up a habit of writing like this because of pylint/codefactor who enforce [no-else-return](https://pylint.pycqa.org/en/latest/user_guide/messages/refactor/no-else-return.html) and personally I prefer it, especially for the final else. Happy to be convinced otherwise.

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    raise ValueError("Unsupported channel dimension format: {}".format(channel_dim))
+
+
+def rescale_image(
```
I prefer the term `rescale`, as we'll also have `normalize`, `rotate`, etc.

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+            Whether or not to apply the scaling factor (to make pixel values integers between 0 and 255). Will default
+            to `True` if the image type is a floating type, `False` otherwise.
+    """
+    if isinstance(image, PIL.Image.Image):
+        return image
+
+    if is_torch_tensor(image) or is_tf_tensor(image):
+        image = image.numpy()
+    elif is_jax_tensor(image):
+        image = np.array(image)
+
+    if not isinstance(image, np.ndarray):
+        raise ValueError("Input image type not supported: {}".format(type(image)))
```
Can't we do if, elif, elif, else raise ValueError here?

amyeroberts replied:
Yep, I'll update

NielsRogge commented on src/transformers/models/glpn/image_processing_glpn.py:
```diff
@@ -0,0 +1,180 @@
+        image = resize(image, (new_h, new_w), resample=resample, data_format=data_format, **kwargs)
+        return image
+
+    def rescale_image(
```
Same here, we could use the term `rescale`.

NielsRogge commented on src/transformers/image_processing_utils.py:
```diff
@@ -0,0 +1,55 @@
+    def __call__(self, images, **kwargs) -> BatchFeature:
+        return self.preprocess(images, **kwargs)
+
+    def preprocess(self, images, **kwargs) -> BatchFeature:
```
Ok, nice :) could possibly also include a `postprocess` method, although I'm wondering whether that could be a method on the model (as it's oftentimes framework-specific)

Definitely stuff for later :) 

cc @alaradirik

amyeroberts replied:
I didn't include a `postprocess` method, as one model can have many different downstream tasks. I was thinking of using a structure similar to the current feature extractors where we have `postprocess_task` methods. What do you think? 

I don't think the postprocessing methods should fall under the model if it's outside of what's needed for training. I wouldn't want to have to load a model in order to process outputs. Re framework specific, we'll have to think about how to handle this as we need to be able to support and have consistent outputs for the different framework implementations. If it's necessary to use specific libraries, we could have the image processor call the specific implementation e.g. `postprocess_instance_segmentation` calls `_postprocess_instance_segmentation_pytorch`.

alaradirik replied:
@amyeroberts @NielsRogge I'm a bit late to the conversation but I agree that postprocessing methods shouldn't fall under the modeling files

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    Returns:
+        image: A converted np.ndarray.
+    """
+    current_channel_dim = infer_channel_dimension_format(image)
```
Would be great to sanity check here that we are dealing with a NumPy array, cause the `infer_channel_dimension_format` assumes that

NielsRogge replied:
Edit: seems like you're dealing with that inside the model-specific image processors

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    Returns:
+        image: A rescaled np.ndarray image.
+    """
+    rescaled_image = image * scale
```
Also here, should we sanity check that we're effectively dealing with a NumPy array?

amyeroberts replied:
Sure! I can add a check. It should probably be added to the other methods too 👍

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    return PIL.Image.fromarray(image)
+
+
+def get_resize_output_image_size(
```
What's the purpose of this method?

amyeroberts replied:
It's not yet because GLPN does its own logic to find the output image size based on `size_divisor`.

NielsRogge commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,246 @@
+    return_numpy: bool = True,
+) -> np.ndarray:
+    """
+    Resizes `image` to (h, w) specified by `size` using the PIL library.
```
Also, won't this method support the same behaviour as torchvision's resize? i.e. when size is an int, it only resizes the smaller edge of the image

amyeroberts replied:
> Given that we use Pillow for resizing, maybe it makes more sense to let size be a tuple of (width, height) rather than the other way around?

As the input and output to the function are numpy arrays, I think that would be confusing. Ideally the user wouldn't have any knowledge about the library used to resize. 

> Also, won't this method support the same behaviour as torchvision's [resize](https://pytorch.org/vision/main/generated/torchvision.transforms.Resize.html)? i.e. when size is an int, it only resizes the smaller edge of the image

I moved the logic to find the dimensions of the output image into `get_resize_output_image_size` so that `resize` does just one thing. This is in part because there's a difference in behaviour between TF and PyTorch. Then each model's image processor can have its own logic within its `resize` method for finding the output shape, or use `get_resize_output_image_size` which has the same behaviour as before.

amyeroberts committed:
Update docs/source/en/internal/image_processing_utils.mdx

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

amyeroberts committed:
Address Niels PR comments

sgugger commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,253 @@
+            (`size`,`size`). If set to `False`, will replicate
+            [`torchvision.transforms.Resize`](https://pytorch.org/vision/stable/transforms.html#torchvision.transforms.Resize)
+            with support for resizing only the smallest edge and providing an optional `max_size`.
+        max_size (`int`, *optional*, defaults to `None`):
```
```suggestion
        max_size (`int`, *optional*):
```
Nit: We don't write the defaults to `None`.

sgugger commented on src/transformers/image_transforms.py:
```diff
@@ -0,0 +1,253 @@
+            The size to use for resizing the image.
+        resample (`int`, *optional*, defaults to `PIL.Image.BILINEAR`):
+            The filter to user for resampling.
+        data_format (`ChannelDimension`, *optional*, defaults to `None`):
```
```suggestion
        data_format (`ChannelDimension`, *optional*):
```

sgugger commented on src/transformers/models/glpn/image_processing_glpn.py:
```diff
@@ -0,0 +1,180 @@
+            resample (`int`, *optional*, defaults to `self.resample`):
+                Resampling filter to use if resizing the image. This can be one of the enum `PIL.Image.Resampling`,
+                Only has an effect if `do_resize` is set to `True`.
+            return_tensors (`str`, *optional*, defaults to `None`):
```
```suggestion
            return_tensors (`str`, *optional*):
```

sgugger reviewed (approved):
Left a couple of nits but looking great!

amyeroberts committed:
Apply suggestions from code review - remove defaults to None

Co-authored-by: Sylvain Gugger <Sylvain.gugger@gmail.com>

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

amyeroberts committed:
Merge branch 'main' into image-transforms-library

amyeroberts commented:
@alaradirik @NielsRogge Could you (re-)review?

alaradirik reviewed (approved):
All looks good to me!

I just have a question regarding multi-modal models such as CLIP and OWL-ViT. These models have both feature extractors and processors, which call their respective tokenizer and feature extractor. Wouldn't creating XXModelProcessor aliases for their feature extractors create issues?

amyeroberts commented:
> I just have a question regarding multi-modal models such as CLIP and OWL-ViT. These models have both feature extractors and processors, which call their respective tokenizer and feature extractor. Wouldn't creating XXModelProcessor aliases for their feature extractors create issues?

@alaradirik I believe this should be OK, as the feature extractors are being mapped to `XxxImageProcessor` rather than `XxxProcessor`, so there's no clash of names. Not sure if this answers your question or I've missed the consequence you're asking about.

amyeroberts committed:
Merge branch 'main' into image-transforms-library

amyeroberts committed:
Fix docstrings and revert to PIL.Image.XXX resampling

Use PIL.Image.XXX resampling values instead of PIL.Image.Resampling.XXX enum as it's only in the recent version >= 9.10 and version is not yet pinned and older version support deprecated

amyeroberts committed:
Some more docstrings and PIL.Image tidy up

amyeroberts committed:
Reorganise arguments so flags by modifiers

amyeroberts committed:
Few last docstring fixes
